crates/dyn-abi/src/ty.rs

```diff
@@ -79,17 +116,89 @@ pub enum DynSolType {
     },
 }
 
-impl core::str::FromStr for DynSolType {
+impl fmt::Display for DynSolType {
+    #[inline]
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.write_str(&self.sol_type_name())
+    }
+}
+
+impl FromStr for DynSolType {
     type Err = DynAbiError;
 
     #[inline]
     fn from_str(s: &str) -> DynAbiResult<Self, Self::Err> {
-        TypeSpecifier::try_from(s).and_then(|t| t.resolve_basic_solidity())
+        Self::parse(s)
     }
 }
 
 impl DynSolType {
+    /// Parses a Solidity type name string into a [`DynSolType`].
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// # use alloy_dyn_abi::DynSolType;
+    /// let type_name = "uint256";
+    /// let ty = DynSolType::parse(type_name)?;
+    /// assert_eq!(ty, DynSolType::Uint(256));
+    /// assert_eq!(ty.sol_type_name(), type_name);
+    /// # Ok::<_, alloy_dyn_abi::DynAbiError>(())
+    /// ```
+    #[inline]
+    pub fn parse(s: &str) -> DynAbiResult<Self> {
+        TypeSpecifier::try_from(s).and_then(|t| t.resolve_basic_solidity())
+    }
+
+    /// Fallible cast to the contents of a variant.
+    #[inline]
+    pub fn as_tuple(&self) -> Option<&[Self]> {
+        match self {
+            Self::Tuple(t) => Some(t),
+            _ => None,
+        }
+    }
+
+    /// Fallible cast to the contents of a variant.
+    #[inline]
+    #[allow(clippy::missing_const_for_fn)]
+    pub fn as_custom_struct(&self) -> Option<(&str, &[String], &[Self])> {
+        match self {
+            #[cfg(feature = "eip712")]
+            Self::CustomStruct {
+                name,
+                prop_names,
+                tuple,
+            } => Some((name, prop_names, tuple)),
+            _ => None,
+        }
+    }
+
+    /// Returns wheter this type is contains a custom struct.
```

Comment:
wheter

should this function only exist if the feature is enabled?

Reply:
I want these to be available always, cfg is already annoying as it is with just one variant